Composer/packages/tools/language-servers/language-generation/src/LGServer.ts

```diff
@@ -0,0 +1,330 @@
+    const document = this.documents.get(params.textDocument.uri);
+    if (!document) {
+      return Promise.resolve(null);
+    }
+    const text = this.getLGDocumentContent(document);
+    const lgResources = getLGResources(text);
+    const templates = lgResources.Templates;
+    const wordRange = getRangeAtPosition(document, params.position);
+    let word = document.getText(wordRange);
+    const matchItem = templates.find(u => u.Name === word);
+    if (matchItem) {
+      const hoveritem: Hover = { contents: [matchItem.Source, matchItem.Body] };
+      return Promise.resolve(hoveritem);
+    }
+    if (word.startsWith('builtin.')) {
+      word = word.substring('builtin.'.length);
+    }
+
+    if (buildInfunctionsMap.has(word)) {
+      const functionEntity = buildInfunctionsMap.get(word);
+      if (!functionEntity) {
+        return Promise.resolve(null);
+      }
+      const hoveritem: Hover = {
+        contents: [
+          `Parameters: ${get(functionEntity, 'Params', []).join(', ')}`,
+          `Documentation: ${get(functionEntity, 'Introduction', '')}`,
+          `ReturnType: ${get(functionEntity, 'Returntype', '').valueOf()}`,
+        ],
+      };
+      return Promise.resolve(hoveritem);
+    }
+    return Promise.resolve(null);
+  }
+
+  protected async resovleSchema(url: string): Promise<string> {
+    const uri = URI.parse(url);
+    if (uri.scheme === 'file') {
+      return new Promise<string>((resolve, reject) => {
+        fs.readFile(uri.fsPath, 'UTF-8', (err, result) => {
+          err ? reject('') : resolve(result.toString());
+        });
+      });
+    }
+    try {
+      const response = await xhr({ url, followRedirects: 5 });
+      return response.responseText;
+    } catch (error) {
+      return Promise.reject(error.responseText || getErrorStatusDescription(error.status) || error.toString());
+    }
+  }
+
+  private removeParamFormat(params: string): string {
+    const paramArr: string[] = params.split(',');
+    const resultArr: string[] = [];
+    paramArr.forEach(element => {
+      resultArr.push(element.trim().split(':')[0]);
+    });
+    return resultArr.join(' ,');
+  }
+
+  private matchedStates(params: TextDocumentPositionParams): { matched: boolean; state: string } | null {
+    const state: string[] = [];
+    const document = this.documents.get(params.textDocument.uri);
+    if (!document) return null;
+    const position = params.position;
+    const range = Range.create(position.line, 0, position.line, position.character);
+    const lineContent = document.getText(range);
+    let flag = false;
+    let finalState = '';
+    if (!lineContent.trim().startsWith('-')) {
+      return { matched: flag, state: finalState };
+    }
+
+    //if the line starts with '-', will try to match
+    flag = true;
+    //initilize the state to plaintext
+    state.push('PlainText');
+    let i = 0;
+    while (i < lineContent.length) {
```

Comment:
please comment as to why the below code is necessary

Reply:
Could you explain?